### PR TITLE
Introduce `toVariableSized` to ConstantSizedArray type

### PR DESCRIPTION
**Pull request by darkdrag00nv2**
Work towards #2530

## Description
Introduce `toVariableSized` function to `ConstantSizedArray` type. This function will return a variable sized array whose contents are copies of the original constant sized array.

Since copy of values are involved, this is not available if the inner type of the constant sized array is a resource.
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/sema/type.go [hunk 1]
```diff
@@ -2025,6 +2025,13 @@ Returns a new array with contents in the reversed order.
 Available if the array element type is not resource-kinded.
 `
 
+const ArrayTypeToVariableSizedFunctionName = "toVariableSized"
+
+const arrayTypeToVariableSizedFunctionDocString = `
+Returns a new variable sized array with the same content.
```
Comment (turbolent) [inline runtime/sema/type.go:2031]:
```suggestion
Returns a new variable-sized array with the copy of the contents of the given array.
```

File: runtime/sema/type.go [hunk 2]
```diff
@@ -2533,6 +2569,16 @@ func ArraySliceFunctionType(elementType Type) *FunctionType {
 	)
 }
 
+func ArrayToVariableSizedFunctionType(elementType Type) *FunctionType {
+	return NewSimpleFunctionType(
+		FunctionPurityImpure,
```
Comment (turbolent) [inline runtime/sema/type.go:2574]: I think it should be possible to make this `view`?

File: runtime/sema/type.go [hunk 3]
```diff
@@ -2533,6 +2569,16 @@ func ArraySliceFunctionType(elementType Type) *FunctionType {
...
+		[]Parameter{},
```
Comment (turbolent) [inline runtime/sema/type.go:2575]: Can this be `nil`?

File: runtime/tests/checker/arrays_dictionaries_test.go [hunk 4]
```diff
@@ -2478,3 +2478,78 @@ func TestCheckDictionaryFunctionEntitlements(t *testing.T) {
...
+	assert.IsType(t, &sema.ExcessiveArgumentsError{}, errs[0])
+}
+
+func TestCheckVariableSizedArrayToVariableSizedInvalid(t *testing.T) {
+
+	t.Parallel()
+
+	_, err := ParseAndCheck(t, `
+		resource X {}
+
```
Comment (turbolent) [inline runtime/tests/checker/arrays_dictionaries_test.go:2523]: Not needed here

```suggestion
```

**Review by turbolent — COMMENTED** [review]
Nice!

File: runtime/sema/type.go (hunk 2 above)
Comment (turbolent) [reply runtime/sema/type.go:2574]: cc @dsainati1

Comment (dsainati1) [reply runtime/sema/type.go:2574]: Yeah, it should be

Comment (darkdrag00nv2) [reply runtime/sema/type.go]: Done, updated.

**Review by turbolent — APPROVED** [review]
Nice!

File: runtime/sema/type.go (hunk 3 above)
Comment (SupunS) [reply runtime/sema/type.go:2575]: I think all those usages are nil-safe. For example, `t.Parameters` is always used in for-loops or with `len(t.Parameters)`, both of which handle `nil` gracefully.
There are already function types with `nil` parameters. e.g:
https://github.com/onflow/cadence/blob/08be35302d32825a01001d8ba217fd32d15e8f51/runtime/sema/runtime_type_constructors.go#L27-L31

```suggestion
		nil,
```

**Review by SupunS — APPROVED** [review]
Look good!

**Pull request merged**